profiling/src/lib.rs

```diff
@@ -339,6 +312,67 @@ pub struct RequestLocals {
     pub vm_interrupt_addr: *const AtomicBool,
 }
 
+/// take a sample every X bytes
+#[cfg(feature = "allocation_profiling")]
+const ALLOCATION_PROFILING_INTERVAL: f32 = 1024.0 * 50.0;
+
+#[cfg(feature = "allocation_profiling")]
+pub struct AllocationProfilingStats {
+    /// number of bytes until next sample collection
+    next_sample: i64,
+}
+
+#[cfg(feature = "allocation_profiling")]
+impl AllocationProfilingStats {
+    fn new() -> AllocationProfilingStats {
+        AllocationProfilingStats {
+            next_sample: AllocationProfilingStats::next_sampling_interval(),
+        }
+    }
+
+    fn next_sampling_interval() -> i64 {
+        Poisson::new(ALLOCATION_PROFILING_INTERVAL)
+            .unwrap()
+            .sample(&mut rand::thread_rng()) as i64
+    }
+
+    fn track_allocation(&mut self, len: u64) {
+        self.next_sample -= len as i64;
+
+        if self.next_sample > 0 {
+            return;
+        }
+
+        let scale = 1.0 / (1.0 - (len as f64 * -1.0 / ALLOCATION_PROFILING_INTERVAL as f64).exp());
+
+        let count = 1.0 * scale;
+        let bytes = len as f64 * scale;
+
+        self.next_sample = AllocationProfilingStats::next_sampling_interval();
```

Comment:
Sampling logic LGTM. This is exactly what the Go profiler does and should work just as well 👍